client/dwh_migration_client/macro_expander.py

```diff
@@ -0,0 +1,249 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Macro handling classes """
+
+import logging
+import re
+from fnmatch import fnmatch
+from typing import Callable, Dict, List, Optional
+
+
+class RecordingLogger:
+    """Logger that keeps track of the messages that it has logged."""
+
+    def __init__(self) -> None:
+        self.messages = []
+
+    def warn_log(self, format_string, *args) -> None:
+        message = format_string.format(*args)
+        self.messages.append(message)
+        logging.warning(message)
+
+
+class MacroExpander(RecordingLogger):
+    """Base class for macro expanders. Do not use this directly."""
```

Comment:
Perhaps make this a first class [ABC](https://docs.python.org/3/library/abc.html). I've never created an ABC that has multiple inheritance so I don't know what gotchas exist there.